contracts/BEP20RewardApeV5.sol

```diff
@@ -0,0 +1,288 @@
+    {
+        STAKE_TOKEN = _stakeToken;
+        REWARD_TOKEN = _rewardToken;
+        rewardPerBlock = _rewardPerBlock;
+        startBlock = _startBlock;
+        bonusEndBlock = _bonusEndBlock;
+
+        // staking pool
+        poolInfo = PoolInfo({
+            lpToken: _stakeToken,
+            allocPoint: 1000,
+            lastRewardBlock: startBlock,
+            accRewardTokenPerShare: 0
+        });
+
+        totalAllocPoint = 1000;
+    }
+
+    // Return reward multiplier over the given _from to _to block.
+    function getMultiplier(uint256 _from, uint256 _to) public view returns (uint256) {
+        if (_to <= bonusEndBlock) {
+            return _to - _from;
+        } else if (_from >= bonusEndBlock) {
+            return 0;
+        } else {
+            return bonusEndBlock - _from;
+        }
+    }
+
+    /// @param  _bonusEndBlock The block when rewards will end
+    function setBonusEndBlock(uint256 _bonusEndBlock) external onlyOwner {
+        require(_bonusEndBlock > block.number, 'new bonus end block must be greater than current');
+        bonusEndBlock = _bonusEndBlock;
+        emit LogUpdatePool(bonusEndBlock, rewardPerBlock);
+    }
+
+    // View function to see pending Reward on frontend.
+    function pendingReward(address _user) external view returns (uint256) {
+        UserInfo storage user = userInfo[_user];
+        uint256 accRewardTokenPerShare = poolInfo.accRewardTokenPerShare;
+        if (block.number > poolInfo.lastRewardBlock && totalStaked != 0) {
+            uint256 multiplier = getMultiplier(poolInfo.lastRewardBlock, block.number);
+            uint256 tokenReward = multiplier * rewardPerBlock * poolInfo.allocPoint / totalAllocPoint;
+            accRewardTokenPerShare = accRewardTokenPerShare + (tokenReward * 1e30 / totalStaked);
+        }
+        return user.amount * accRewardTokenPerShare / 1e30 - user.rewardDebt;
+    }
+
+    // Update reward variables of the given pool to be up-to-date.
+    function updatePool() public {
+        if (block.number <= poolInfo.lastRewardBlock) {
+            return;
+        }
+        if (totalStaked == 0) {
+            poolInfo.lastRewardBlock = block.number;
+            return;
+        }
+        uint256 multiplier = getMultiplier(poolInfo.lastRewardBlock, block.number);
+        uint256 tokenReward = multiplier * rewardPerBlock * poolInfo.allocPoint / totalAllocPoint;
+        totalRewardsAllocated += tokenReward;
+        poolInfo.accRewardTokenPerShare = poolInfo.accRewardTokenPerShare + (tokenReward * 1e30 / totalStaked);
+        poolInfo.lastRewardBlock = block.number;
+    }
+
+
+    /// Deposit staking token into the contract to earn rewards.
+    /// @dev Since this contract needs to be supplied with rewards we are
+    ///  sending the balance of the contract if the pending rewards are higher
+    /// @param _amount The amount of staking tokens to deposit
+    function deposit(uint256 _amount, address _user) external nonReentrant {
```

Comment:
Recommendation for backward compatibility 
- [ ] change this function to `internal` `function deposit(uint256 _amount, address _to) internal nonReentrant `
- [ ] leave the old `deposit(uint256 _amount) external nonReentrant` for backward compatbility
- [ ] add `depositTo(uint256 _amount, _address _to) external nonReentrant`